django/contrib/messages/test.py

```diff
@@ -0,0 +1,26 @@
+from .api import get_messages
+from .utils import get_level_tags
+
+
+class MessagesTestMixin:
+    def assertMessages(self, response, expected_messages):
+        level_to_name = get_level_tags()
+
+        def level_name(level):
+            level_name = level_to_name.get(level, level)
+            if isinstance(level_name, str) and level_name:
+                return level_name.upper()
+            return level
+
+        request_messages = []
+        for message in get_messages(response.wsgi_request):
+            actual = (level_name(message.level), message.message, message.extra_tags)
+            request_messages.append(actual)
+        expected_msgs = []
+        for message in expected_messages:
+            level = message[0]
+            expected = (level_name(level), *message[1:])
+            if len(expected) == 2:
+                expected = expected + ("",)  # The default for extra_tags.
+            expected_msgs.append(expected)
+        self.assertEqual(request_messages, expected_msgs)
```

Comment:
What do you think about passing `Message` objects in `expected_messages` instead of comparing tuples? or changing `expected_messages` to the list of `Message` objects before calling `assertEqual()`? For example:
```python
def get_tag_levels():
    return {
        **constants.DEFAULT_LEVELS,
        **{v:k for k, v in getattr(settings, "MESSAGE_TAGS", {}).items()},
    }

        
class MessagesTestMixin:
    def assertMessages(self, response, expected_messages):
        name_to_level = get_tag_levels()
        expected_msgs = []
        for message in expected_messages:
            level = message[0]
            if isinstance(level, str):
                level = name_to_level.get(level, level)
            expected_msgs.append(Message(level, *message[1:]))
        self.assertEqual(get_messages(response.wsgi_request), expected_msgs)

```

Reply:
Messages are shown with `__repr__` in the failure output, which looks like `<django.contrib.messages.storage.base.Message object at 0x61ebe92d3e50>`, so the assertion failure isn’t as helpful:
```python
# --- test
def test_failure_output(self):
    self.assertEqual(
        base.Message(constants.INFO, "Message"),
        base.Message(constants.WARNING, "Message"),
    )
# --- output
======================================================================
FAIL: test_failure_output (messages_tests.tests.AssertMessagesTest.test_level_without_name)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/freitafr/dev/django/django/test/utils.py", line 443, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/freitafr/dev/django/tests/messages_tests/tests.py", line 124, in test_level_without_name
    self.assertEqual(
AssertionError: <django.contrib.messages.storage.base.Message object at 0x69a6956abf10> != <django.contrib.messages.storage.base.Message object at 0x69a6956dc1d0>
```

It also requires importing `Message`, which is a bit cumbersome.


---
Maybe we should accept `Message`s for expected_messages, and convert them to tuples?

Reply:
[`test_mismatching_length`](https://github.com/django/django/pull/17101/files#diff-37c7854e7c0cdd5bc116efcbc1f2dbc043debac22c4f1471f9a190c427f5b424R137) gives an example output of the failure message in case of tuples.

Reply:
> Messages are shown with `__repr__` in the failure output, which looks like `<django.contrib.messages.storage.base.Message object at 0x61ebe92d3e50>`, so the assertion failure isn’t as helpful:

We can always import `Message.__repr__`.

Reply:
You probably meant implement?

It will not be as helpful when producing failures, as we’ll be diff-ing strings. The tuples have more meaning to them (level, message, tags). But a carefully crafted `__repr__` might be good enough?

Reply:
Using `Message` makes `assertMessages` much simpler (IMO) so I'd try to implement `__repr__`, but maybe comparing tuples is fine :shrug:

Reply:
> Having a more useful `__repr__` for `Message` might be a win in any case.

Yes, we can add it separately with `Refs #24121 -- Added __repr__() to Message.`.

Reply:
https://github.com/django/django/pull/17188